### ISSUE TITLE
Emdedding computation is too slow

### ISSUE THREAD

**Issue by atharvadeore999** [QwenLM/Qwen3-Embedding#60]
I am using the following code to extract the embeddings.The data has the shape of  (1436116, 1).The model utilizes GPU but still it takes alot of time to compute embeddings.Please help to solve the problem:
import pandas as pd
import torch
import gc
from tqdm import tqdm
from sentence_transformers import SentenceTransformer

# Check GPU availability
print(f"CUDA available: {torch.cuda.is_available()}")
print(f"CUDA device count: {torch.cuda.device_count()}")
if torch.cuda.is_available():
    print(f"Current CUDA device: {torch.cuda.current_device()}")
    print(f"CUDA device name: {torch.cuda.get_device_name()}")

# Load data
unique_products = pd.read_csv("/home/atharva/notebooks/cleaned_unique_products.csv")
unique_products = unique_products[unique_products['product_text'].str.len() > 0].copy()

# Extract product texts (this was missing in your original code)
product_texts = unique_products['product_text'].tolist()

print(f"Total products to process: {len(product_texts)}")

# Load model and move to GPU
model_name = "Qwen/Qwen3-Embedding-0.6B"  # Replace with your actual model name
device = 'cuda' if torch.cuda.is_available() else 'cpu'
print(f"Using device: {device}")

# Load model with GPU support
model = SentenceTransformer(model_name, device=device)

# Alternative way to ensure model is on GPU
# model = SentenceTransformer(model_name)
# model = model.to(device)

print(f"Model device: {next(model.parameters()).device}")

# Batch processing parameters
batch_size = 64  # You might want to increase this for GPU (try 128, 256, or 512)
total_batches = (len(product_texts) + batch_size - 1) // batch_size

print(f"Processing {len(product_texts)} texts in {total_batches} batches of size {batch_size}")

# Process embeddings
embeddings_list = []

for i in tqdm(range(0, len(product_texts), batch_size), total=total_batches, desc="Processing Batches"):
    batch_texts = product_texts[i:i + batch_size]
    
    with torch.no_grad():
        # The model.encode() should automatically use GPU if model is on GPU
        batch_embeddings = model.encode(
            batch_texts, 
            show_progress_bar=False,
            convert_to_numpy=True,  # This helps with memory management
            normalize_embeddings=False  # Set to True if you want normalized embeddings
        )
    
    # Convert to list and append
    for embedding in batch_embeddings:
        embeddings_list.append(embedding.tolist())
    
    # Memory cleanup
    del batch_embeddings
    if torch.cuda.is_available():
        torch.cuda.empty_cache()
    gc.collect()
    
    # Optional: Print progress every 100 batches
    if (i // batch_size + 1) % 100 == 0:
        print(f"Processed {i // batch_size + 1}/{total_batches} batches")

print(f"Generated {len(embeddings_list)} embeddings")
